### ISSUE TITLE
Parallel tasks call "finally" function multiple times  (a call for every error)

### ISSUE THREAD

**Issue by triadium** [mynameisdaniil/yaff#3]
Parallel tasks call **finally** function multiple times (a call for every error), but callback already done for a first error - app state changed.
I think better to call **finally** function with an array of errors. For a **seq** function set an "one-item" array of errors, for a **par** function put error to its position of calling. And **null** unless no errors.

**Comment by mynameisdaniil** [comment]
Well, all I can do is to call cb once for the first error. Idea with array won't work because of this two things:
1) This is not a nodejs style of calling callbacks, so composability would be broken
2) I can't really say how much errors to expect, because async functions can call *this* many times

**Comment by triadium** [comment]
Our task:
Files are generated in a number of parallel processes that use the same source of format specification in our project. It was decided to replace the library "Seq" to "Yaff", as do not like the behavior of the first one in "catch" situations.
The main problem is that the parallel processes are independent of each other and while their performance is not known beforehand. But each concurrent process has its place in the final result array and the number of parallel processes is known. 
Accordingly, one could allocate an array for errors on the same principle - ie wait all the results or errors that will be at their positions in the respective arrays.

1) I agree, but you can make a special class of errors, which would contain an array of descriptions of errors. The main goal for me to inform to the client all the errors on the processes that run in parallel, and not to do output of the description one by one at each fix and restart.
2) a call *this* many times in the same async function context  has side effects  and, therefore, we don't use such approach anywhere, since the main task of the library for us - to return the result and put it in the exact location.

Is it possible to make such changes on the basis of these assumptions?
I would like to use the code of the "master" branch always.

**Comment by mynameisdaniil** [comment]
Well, that makes sense and while this is not the intended way of handling
errors it is not too hard to add something like *.collectAll()* to collect
all errors before *.finally()* gets called.
30 ÍÁÒÔÁ 2015 Ç. 13:23 ÐÏÌØÚÏ×ÁÔÅÌØ "triadium" <notifications@github.com>
ÎÁÐÉÓÁÌ:

**Comment by triadium** [comment]
```javascript
YAFF([in1, in2, in3, ...])
.parMap(...)    // assume that errors have occurred in some of the processes...
.unflatten()     // skip it
.seq(...)          // skip it
.flatten(false) // skip it
.parMap(...)   // skip it
.collectAll( /* like unflatten or like seq with callback argument? */ )
.finally(function(errorOfSpecialClass, results){
    errorOfSpecialClass === {message, errors: [null, Error, Error, null, ...], ...};
    // It is desirable to be able to process the results of some of the processes, 
    // completed without errors in this function, even when errors occured 
    // in some of the other parallel processes.
    results === [value1, null, null, value4, ...]; 
  
   // for an error in a seq function  
   // errorOfSpecialClass === {message, errors: [Error], ...}
   // results === [null]; // or just null
});
```
Is that way?

Will *.collectAll()* function collect only errors or also results like in the above example?
Names of functions are very important for us.
We have got small chains usually, and track the status and restore processing for positive results from the first step, missed because of an error, not so difficult and saves a lot of time for large volumes of data to be processed.

 
Show your vision as an example, please.

**Comment by mynameisdaniil** [comment]
I think it will work like `unflatten()` e.g.
```javascript
var myAsyncFun = function (images, cb) {
  YAFF(images)
    .parMap(...) //will return array of all errors into cb
    .unfaltten()
    .seq(...) //will return single error
    .collectAll()
    .finally(cb)
};

In case of errors in `parMap(...)` it will return array of errors into cb, in case of error in `seq(..)` it will return single error. But let's see how everything will turn out. I'm working on implementation right now.

```

**Comment by mynameisdaniil** [comment]
Well, it turned out that you don't need `collectAll`. I decided to collect errors produced by groups of parallel functions into array of errors. (in other words just remove `collectAll` from my previous example)
Code requires more testing and optimisation, so I won't push it to npm for now. Please, check out master branch, take a look and say what you think.

**Comment by triadium** [comment]
Works great! Situation "Unexpected errors" is processed as expected. 
I thought that the results obtained in the previous steps can be saved with the current state of the chain in the external context, so this functionality is sufficient for appropriate behavior in all situations.
Thanks!

**Comment by mynameisdaniil** [comment]
published: 0.1.0@npmjs/yaff

**Comment by mynameisdaniil** [comment]
BTW I recommend updating to latest (1.02) version since this version correctly handles error arrays.

**Comment by triadium** [comment]
Not the right level of structure for err argument. 
0.1.2  - with **[e]**
[ [ [ { [Error: ENOENT, open 'D:\projects\template.tmpl']
        errno: 34,
        code: 'ENOENT',
        path: 'D:\\projects\\template.tmpl' } ],
    null ] ]
	
Before 0.1.2  - with **e**
[ { [Error: ENOENT, open 'D:\projects\template.tmpl']
    errno: 34,
    code: 'ENOENT',
    path: 'D:\\projects\\template.tmpl' } ]

**Comment by mynameisdaniil** [comment]
Sorry for late answer. Better to open new issue if you think behaviour is wrong. Not sure I understood the problem.
